### PR TITLE
ARROW-9861: [Java] Support big-endian in DecimalVector

### PR DESCRIPTION
**Pull request by kiszk**
This PR fixes failures in DecimalVectorTest on a big-endian platform

### CODE REVIEW

**Comment by emkornfield** [comment]
@kiszk you'll probably need to make some fixes for Decimal256 as well, I think.

@BryanCutler you mentioned you could devote some time to reviewing Big Endian changes?  Would you mind taking a look through this one and  @kiszk other Java changes?

**Comment by BryanCutler** [comment]
Sure, I can take a look. It might a day or two before I can though @kiszk .

**Comment by kiszk** [comment]
@emkornfield Sure, I will work for supporting Decimal256.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:171]: you could move `padBytes` out of the `if` statements

File: java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java:49]: I don't think you need 2 loops here, just move the `Integer.MAX_VALUE` and `MIN_VALUE` here

**Review by BryanCutler — CHANGES_REQUESTED** [review]
LGTM, just a couple minor things

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java
Comment (kiszk) [reply java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:141]: This is used for `DecialVector` that is a fixed-width (16-byte) vector. This routine extends the signed bit to a new long. I will write a document in the comment.

File: java/vector/src/main/java/org/apache/arrow/vector/DecimalVector.java
Comment (kiszk) [reply java/vector/src/main/java/org/apache/arrow/vector/DecimalVector.java:244]: If we add `if (length == TYPE_WIDTH) return` before line 244, no data is copied from `value` to `outAddress`. I will add a comment `copy data from value to outAddress` after line 244.

File: java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java
Comment (kiszk) [reply java/vector/src/test/java/org/apache/arrow/vector/util/DecimalUtilityTest.java:49]: Sure, I will drop lines 49-57.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java
Comment (kiszk) [reply java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:141]: In addition, this PR will support to write it to a 256-bit entry for the future.

**Comment by kiszk** [comment]
I will update the Decimal256Vector class late today or tomorrow.

File: java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java
Comment (BryanCutler) [inline java/vector/src/main/java/org/apache/arrow/vector/util/DecimalUtility.java:138]: typo: `DeciimalUtility` -> `DecimalUtility`

File: java/vector/src/test/java/org/apache/arrow/vector/complex/impl/TestComplexCopier.java
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/impl/TestComplexCopier.java:529]: Could you use `DecimalVector.TYPE_WIDTH` here?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java:313]: Same here

**Review by BryanCutler — COMMENTED** [review]
Looks good, just a few minor nits

**Review by BryanCutler — APPROVED** [review]
LGTM. Looks like a flaky test, will try again.

**Comment by BryanCutler** [comment]
The test error with Java JNI looks unrelated and seems to be an env issue with ORC, I'll go ahead with merging this.

**Comment by BryanCutler** [comment]
merged to master, thanks @kiszk !

**Comment by kiszk** [comment]
@BryanCutler Thank you. One comment.   
According to [the document](https://github.com/apache/arrow/blob/master/docs/source/developers/contributing.rst#endianness), benchmarking is necessary before merging features (I think that the test code does not affect the performance)
> Benchmarks for performance critical parts of the code to demonstrate no regression.

I am working for benchmark bot for Java [here](https://github.com/apache/arrow/pull/8210). It would be good to merge new features after this bot will be available.

cc @emkornfield

**Comment by BryanCutler** [comment]
I did not see anything that looks like it would affect performance here, but I agree we should get some benchmarks going to be sure. I will look at you other PR next.

**Pull request closed without merging**
